File: elixir/triangle/triangle.exs
defmodule Triangle do
  @type kind :: :equilateral | :isosceles | :scalene

  @doc """
  Return the kind of triangle of a triangle with 'a', 'b' and 'c' as lengths.
  """
  @spec kind(number, number, number) :: { :ok, kind } | { :error, String.t }
  def kind(a, b, c) do
    cond do
      negative_length(a, b, c) -> {:error, "all side lengths must be positive"}
      violate_triangle_equality(a, b, c) -> {:error, "side lengths violate triangle inequality"}
      is_scalene(a, b, c) -> {:ok, :scalene}
      is_equilateral(a, b, c) -> {:ok, :equilateral}
      is_isosceles(a, b, c) -> {:ok, :isosceles}
      true -> {:error, "Not a triangle"}
    end
  end

  def negative_length(a, b, c) do
    a <= 0 or b <= 0 or c <= 0
  end

  def violate_triangle_equality(a, b, c) do
    a + b <= c or b + c <= a or a + c <= b
  end

  def is_scalene(a, b, c) do
    a !== b and b !== c and a !== c
  end

  def is_equilateral(a, b, c) do
    a === b and b === c
  end

  def is_isosceles(a, b, c) do
    cond do
      a === b and a !== c -> true
      b === c and a !== b -> true
      a === c and a !== b -> true
      true -> false
    end
  end
end

File: go/triangle/triangle.go
package triangle

type Kind string

const (
  NaT = "NotTriangle"
  Equ = "Equilateral"
  Iso = "Isosceles"
  Sca = "Scalene"
)

// KindFromSides should have a comment documenting it.
func KindFromSides(a, b, c float64) Kind {
    if !isTriangle(a, b, c) {
        return NaT
    }
    if isEquilateral(a, b, c) {
        return Equ
    }
    if isIsosceles(a, b, c) {
        return Iso
    }
    return Sca
}

func isTriangle(a, b, c float64) bool {
    zeroLengthSide := a <= 0 || b <= 0 || c <= 0
    return !zeroLengthSide
}

func isEquilateral(a, b, c float64) bool {
    return a == b && b == c
}

func isIsosceles(a, b, c float64) bool {
    return a == b || a == c || b == c
}

func isScalene(a, b, c float64) bool {
    return a != b && b != c && a != c
}
